Add support for Background Fetch persistent storage

youennf opened the pull request:
#### 5a9caebbb042cd1aebc44514f8a6bdf036f3478b
<pre>
Add support for Background Fetch persistent storage
<a href="https://bugs.webkit.org/show_bug.cgi?id=252831">https://bugs.webkit.org/show_bug.cgi?id=252831</a>
rdar://problem/105835505

Reviewed by Sihui Liu.

Implement BackgroundFetchStoreImpl by introducing BackgroundFetchStoreManager which is responsible for reading and writing files.
BackgroundFetchStoreManager runs in NetworkStorageManager work queue and is created by OriginStorageManager.
We introduce a helper routine to get a BackgroundFetchStoreManager from its NetworkStorageManager from main thread via a Function.
We introduce a way to push stored background fetches to BackgroundFetchEngine when loading a given BackgroundFetchStoreManager.

Covered by LayoutTests/http/wpt/background-fetch/background-fetch-persistency.window.html.

* LayoutTests/http/wpt/background-fetch/background-fetch-persistency.window-expected.txt: Added.
* LayoutTests/http/wpt/background-fetch/background-fetch-persistency.window.html: Added.
* LayoutTests/http/wpt/background-fetch/background-fetch-persistency.window.js: Added.
(promise_test.async t):
* LayoutTests/http/wpt/background-fetch/resources/trickle.py: Added.
(main):
* LayoutTests/http/wpt/background-fetch/sw.js:
* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/Headers.cmake:
* Source/WebCore/platform/WebCorePersistentCoders.cpp:
(WTF::Persistence::Coder&lt;WebCore::ImageResource&gt;::encode):
(WTF::Persistence::Coder&lt;WebCore::ImageResource&gt;::decode):
* Source/WebCore/platform/WebCorePersistentCoders.h:
* Source/WebCore/workers/service/ServiceWorkerRegistrationKey.cpp:
(WebCore::ServiceWorkerRegistrationKey::clientOrigin const):
* Source/WebCore/workers/service/ServiceWorkerRegistrationKey.h:
* Source/WebCore/workers/service/background-fetch/BackgroundFetch.cpp:
(WebCore::BackgroundFetch::BackgroundFetch):
(WebCore::BackgroundFetch::didFinishRecord):
(WebCore::BackgroundFetch::recordIsCompleted):
(WebCore::BackgroundFetch::Record::didFinish):
(WebCore::BackgroundFetch::doStore):
(WebCore::BackgroundFetch::createFromStore):
* Source/WebCore/workers/service/background-fetch/BackgroundFetch.h:
(WebCore::BackgroundFetch::registrationKey const):
(WebCore::BackgroundFetch::setRecords):
* Source/WebCore/workers/service/background-fetch/BackgroundFetchEngine.cpp:
(WebCore::BackgroundFetchEngine::addFetchFromStore):
* Source/WebCore/workers/service/background-fetch/BackgroundFetchEngine.h:
* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::NetworkSession::createBackgroundFetchStore):
* Source/WebKit/NetworkProcess/storage/BackgroundFetchStoreImpl.cpp:
(WebKit::BackgroundFetchStoreImpl::BackgroundFetchStoreImpl):
(WebKit::BackgroundFetchStoreImpl::getFilename):
(WebKit::BackgroundFetchStoreImpl::initializeFetches):
(WebKit::BackgroundFetchStoreImpl::clearFetch):
(WebKit::BackgroundFetchStoreImpl::clearAllFetches):
(WebKit::BackgroundFetchStoreImpl::storeFetch):
(WebKit::BackgroundFetchStoreImpl::storeFetchResponseBodyChunk):
(WebKit::BackgroundFetchStoreImpl::retrieveResponseBody):
* Source/WebKit/NetworkProcess/storage/BackgroundFetchStoreImpl.h:
(WebKit::BackgroundFetchStoreImpl::create):
* Source/WebKit/NetworkProcess/storage/BackgroundFetchStoreManager.cpp: Added.
(WebKit::shouldUseFileMapping):
(WebKit::BackgroundFetchStoreManager::createNewStorageIdentifier):
(WebKit::BackgroundFetchStoreManager::BackgroundFetchStoreManager):
(WebKit::BackgroundFetchStoreManager::~BackgroundFetchStoreManager):
(WebKit::BackgroundFetchStoreManager::initializeFetches):
(WebKit::BackgroundFetchStoreManager::clearFetch):
(WebKit::BackgroundFetchStoreManager::clearAllFetches):
(WebKit::BackgroundFetchStoreManager::storeFetch):
(WebKit::createFetchResponseBodyFile):
(WebKit::BackgroundFetchStoreManager::storeFetchResponseBodyChunk):
(WebKit::BackgroundFetchStoreManager::retrieveResponseBody):
* Source/WebKit/NetworkProcess/storage/BackgroundFetchStoreManager.h: Added.
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::dispatchTaskToBackgroundFetchManager):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/NetworkProcess/storage/OriginStorageManager.cpp:
(WebKit::OriginStorageManager::StorageBucket::toStorageType const):
(WebKit::OriginStorageManager::StorageBucket::toStorageIdentifier const):
(WebKit::OriginStorageManager::StorageBucket::backgroundFetchManager):
(WebKit::OriginStorageManager::StorageBucket::resolvedBackgroundFetchStoragePath):
(WebKit::OriginStorageManager::StorageBucket::resolvedPath):
(WebKit::OriginStorageManager::backgroundFetchManager):
* Source/WebKit/NetworkProcess/storage/OriginStorageManager.h:
* Source/WebKit/Shared/WebsiteData/WebsiteData.cpp:
(WebKit::WebsiteData::ownerProcess):
* Source/WebKit/Shared/WebsiteData/WebsiteDataType.h:
* Source/WebKit/Shared/WebsiteData/WebsiteDataType.serialization.in:
* Source/WebKit/Sources.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/260990@main">https://commits.webkit.org/260990@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/42755b524a333372e4ffc620479ae0d3508f4f9d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/110205 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/19302 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/42863 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/1624 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/119194 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/114156 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/20763 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/10493 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/102463 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/115950 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/15457 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/98661 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/43673 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/97418 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/30325 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/85516 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/11992 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/31662 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/12604 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/8629 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/17970 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/51279 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7612 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/14413 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

szewai reviewed (commented):
Some initial thoughts:
1. I think BackgroundFetchStoreBackend should not have direct access to members (`WorkQueue`, `OriginStorageManager`, etc) of `NetworkStorageManager`; it should invoke `NetworkStorageManager` methods (e.g. `NetworkStorageManager::backgroundFetchStore`). `NetworkStorageManager` is self-contained now, and knows to perform adequate checks (e.g. assert it is not closed when it receives a call from main thread).  We might achieve the same thing by expose more states (like `m_closed`) of `NetworkStorageManager` so `BackgroundFetchStoreBackend` can check, but I don't think that's essentially better than current pattern. 
2. `BackgroundFetchStoreBackend` might be renamed to `BackgroundFetchStorageManager` to be compatible with other types (we use `TypeStorageManager` for consistency)
3. Should we move the data encoding and decoding to `BackgroundFetchStoreBackend`, so they do not block main thread and `BackgroundFetchStore` could update its map in the callback for `initalizeFetches`?
4. If BackgroundFetch code needs to perform quota check, `QuotaManager` should calculate the size of BackgroundFetchStorage in its usage

szewai commented on Source/WebKit/Shared/WebsiteData/WebsiteDataType.h:
```diff
@@ -51,6 +51,9 @@ enum class WebsiteDataType : uint32_t {
 #endif
     FileSystem = 1 << 19,
+#if ENABLE(SERVICE_WORKER)
+    BackgroundFetchStorage = 1 << 20,
```
I think you need to update WebsiteDataType.serialization.in too

youennf replied:
ok

szewai commented on Source/WebKit/NetworkProcess/storage/BackgroundFetchStoreManager.cpp:
```diff
@@ -0,0 +1,217 @@
+            FileSystem::deleteFile(filePath);
+
+            size_t index = 0;
+            String bodyPath;
+            do {
+                bodyPath = makeString(filePath, "-", index++);
+            } while (FileSystem::deleteFile(bodyPath));
```
Does this mean if we fail to delete one file, we won't delete the rest? Seems not ideal.
Probably we should iterate all files and delete the ones that have the prefix in name

youennf replied:
Will change

szewai commented on Source/WebKit/NetworkProcess/storage/BackgroundFetchStoreManager.cpp:
```diff
@@ -0,0 +1,217 @@
+
+static constexpr auto fetchSuffix = "-backgroundfetch"_s;
+
+static bool shouldUseFileMapping(uint64_t fileSize)
```
It seems we could make this return `sharedbuffer::MayUseFileMapping` directly

youennf replied:
Returning a bool is easier to read.

szewai commented on Source/WebKit/NetworkProcess/storage/BackgroundFetchStoreManager.cpp:
```diff
@@ -0,0 +1,217 @@
+    m_ioQueue->dispatch([queue = Ref { m_taskQueue }, filePath = WTFMove(filePath).isolatedCopy(), data = Ref { data }, callback = WTFMove(callback)]() mutable {
+        auto result = StoreResult::InternalError;
+        FileSystem::PlatformFileHandle handle = FileSystem::openFile(filePath, FileSystem::FileOpenMode::ReadWrite);
+        if (handle) {
```
`if (FileSystem::isHandleValid(handle))`

youennf replied:
ok

szewai commented on Source/WebKit/NetworkProcess/storage/BackgroundFetchStoreManager.cpp:
```diff
@@ -0,0 +1,217 @@
+    }
+
+    auto chunkIdentifier = makeString(identifier, "-", index);
+    auto filePath = FileSystem::pathByAppendingComponents(m_path, { chunkIdentifier });
```
`pathByAppendingComponent`

youennf replied:
ok

szewai commented on Source/WebKit/NetworkProcess/storage/BackgroundFetchStoreManager.cpp:
```diff
@@ -0,0 +1,217 @@
+    auto filePath = FileSystem::pathByAppendingComponents(m_path, { identifier });
+    m_ioQueue->dispatch([queue = Ref { m_taskQueue }, filePath = WTFMove(filePath).isolatedCopy(), data = WTFMove(data), callback = WTFMove(callback)]() mutable {
+        auto writeResult = FileSystem::overwriteEntireFile(filePath, { data.data(), data.size() });
+        auto result = writeResult != -1 ? StoreResult::OK : StoreResult::InternalError;
```
Should write size < buffer size be counted as failure?

youennf replied:
Probably, I'll add a FIXME like done below for body chunks

szewai commented on Source/WebKit/NetworkProcess/storage/BackgroundFetchStoreManager.cpp:
```diff
@@ -0,0 +1,217 @@
+        return;
+    }
+
+    auto chunkIdentifier = makeString(identifier, "-", index);
```
Seem we should make a function to generate chunk file name

youennf replied:
ok

szewai commented on LayoutTests/http/wpt/background-fetch/background-fetch-persistency.window.js:
```diff
@@ -0,0 +1,40 @@
+  assert_equals(ids.length, 2, "v2");
+  assert_true(uniqueId1 == ids[0] || uniqueId1 == ids[1], "uniqueId1 v2");
+  assert_true(uniqueId2 == ids[0] || uniqueId2 == ids[1], "uniqueId2 v2");
+}, "fetch should fail with NotAllowedError if permission is denied");
```
Where is `NotAllowedError`?

szewai commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -1321,5 +1323,17 @@ void NetworkStorageManager::cacheStorageRepresentation(CompletionHandler<void(St
 }
 
+#if ENABLE(SERVICE_WORKER)
+void NetworkStorageManager::dispatchTaskToBackgroundFetchManager(const WebCore::ClientOrigin& origin, Function<void(BackgroundFetchStoreManager&)>&& callback)
```
`NetworkStorageManager` should not accept new task on its queue when `m_closed` is true (it dispatches a cleanup task to the queue when close, which is supposed to be the last task).
I would prefer adding individual functions like `NetworkStorageManager::backgroundFetchClearFetch` to not expose `BackgroundFetchManager`  to other class (as technically they will be able to capture it). 

We may do that as followup and return `BackgroundFetchStoreManager*` here.

youennf replied:
I'll add a `if(m_closed)` check.
I'll also change BackgroundFetchStoreManager to not be ref counted.

szewai reviewed (approved):
r=me with comments

youennf committed:
Add support for Background Fetch persistent storage
https://bugs.webkit.org/show_bug.cgi?id=252831
rdar://problem/105835505

Reviewed by Sihui Liu.

Implement BackgroundFetchStoreImpl by introducing BackgroundFetchStoreManager which is responsible for reading and writing files.
BackgroundFetchStoreManager runs in NetworkStorageManager work queue and is created by OriginStorageManager.
We introduce a helper routine to get a BackgroundFetchStoreManager from its NetworkStorageManager from main thread via a Function.
We introduce a way to push stored background fetches to BackgroundFetchEngine when loading a given BackgroundFetchStoreManager.

Covered by LayoutTests/http/wpt/background-fetch/background-fetch-persistency.window.html.

* LayoutTests/http/wpt/background-fetch/background-fetch-persistency.window-expected.txt: Added.
* LayoutTests/http/wpt/background-fetch/background-fetch-persistency.window.html: Added.
* LayoutTests/http/wpt/background-fetch/background-fetch-persistency.window.js: Added.
(promise_test.async t):
* LayoutTests/http/wpt/background-fetch/resources/trickle.py: Added.
(main):
* LayoutTests/http/wpt/background-fetch/sw.js:
* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/Headers.cmake:
* Source/WebCore/platform/WebCorePersistentCoders.cpp:
(WTF::Persistence::Coder<WebCore::ImageResource>::encode):
(WTF::Persistence::Coder<WebCore::ImageResource>::decode):
* Source/WebCore/platform/WebCorePersistentCoders.h:
* Source/WebCore/workers/service/ServiceWorkerRegistrationKey.cpp:
(WebCore::ServiceWorkerRegistrationKey::clientOrigin const):
* Source/WebCore/workers/service/ServiceWorkerRegistrationKey.h:
* Source/WebCore/workers/service/background-fetch/BackgroundFetch.cpp:
(WebCore::BackgroundFetch::BackgroundFetch):
(WebCore::BackgroundFetch::didFinishRecord):
(WebCore::BackgroundFetch::recordIsCompleted):
(WebCore::BackgroundFetch::Record::didFinish):
(WebCore::BackgroundFetch::doStore):
(WebCore::BackgroundFetch::createFromStore):
* Source/WebCore/workers/service/background-fetch/BackgroundFetch.h:
(WebCore::BackgroundFetch::registrationKey const):
(WebCore::BackgroundFetch::setRecords):
* Source/WebCore/workers/service/background-fetch/BackgroundFetchEngine.cpp:
(WebCore::BackgroundFetchEngine::addFetchFromStore):
* Source/WebCore/workers/service/background-fetch/BackgroundFetchEngine.h:
* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::NetworkSession::createBackgroundFetchStore):
* Source/WebKit/NetworkProcess/storage/BackgroundFetchStoreImpl.cpp:
(WebKit::BackgroundFetchStoreImpl::BackgroundFetchStoreImpl):
(WebKit::BackgroundFetchStoreImpl::getFilename):
(WebKit::BackgroundFetchStoreImpl::initializeFetches):
(WebKit::BackgroundFetchStoreImpl::clearFetch):
(WebKit::BackgroundFetchStoreImpl::clearAllFetches):
(WebKit::BackgroundFetchStoreImpl::storeFetch):
(WebKit::BackgroundFetchStoreImpl::storeFetchResponseBodyChunk):
(WebKit::BackgroundFetchStoreImpl::retrieveResponseBody):
* Source/WebKit/NetworkProcess/storage/BackgroundFetchStoreImpl.h:
(WebKit::BackgroundFetchStoreImpl::create):
* Source/WebKit/NetworkProcess/storage/BackgroundFetchStoreManager.cpp: Added.
(WebKit::shouldUseFileMapping):
(WebKit::BackgroundFetchStoreManager::createNewStorageIdentifier):
(WebKit::BackgroundFetchStoreManager::BackgroundFetchStoreManager):
(WebKit::BackgroundFetchStoreManager::~BackgroundFetchStoreManager):
(WebKit::BackgroundFetchStoreManager::initializeFetches):
(WebKit::BackgroundFetchStoreManager::clearFetch):
(WebKit::BackgroundFetchStoreManager::clearAllFetches):
(WebKit::BackgroundFetchStoreManager::storeFetch):
(WebKit::createFetchResponseBodyFile):
(WebKit::BackgroundFetchStoreManager::storeFetchResponseBodyChunk):
(WebKit::BackgroundFetchStoreManager::retrieveResponseBody):
* Source/WebKit/NetworkProcess/storage/BackgroundFetchStoreManager.h: Added.
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::dispatchTaskToBackgroundFetchManager):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/NetworkProcess/storage/OriginStorageManager.cpp:
(WebKit::OriginStorageManager::StorageBucket::toStorageType const):
(WebKit::OriginStorageManager::StorageBucket::toStorageIdentifier const):
(WebKit::OriginStorageManager::StorageBucket::backgroundFetchManager):
(WebKit::OriginStorageManager::StorageBucket::resolvedBackgroundFetchStoragePath):
(WebKit::OriginStorageManager::StorageBucket::resolvedPath):
(WebKit::OriginStorageManager::backgroundFetchManager):
* Source/WebKit/NetworkProcess/storage/OriginStorageManager.h:
* Source/WebKit/Shared/WebsiteData/WebsiteData.cpp:
(WebKit::WebsiteData::ownerProcess):
* Source/WebKit/Shared/WebsiteData/WebsiteDataType.h:
* Source/WebKit/Shared/WebsiteData/WebsiteDataType.serialization.in:
* Source/WebKit/Sources.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/260990@main
